srml/grandpa/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+				auth_count.encode(),
+			);
+		});
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		/// Report some misbehaviour.
+		fn report_misbehavior(origin, _report: Vec<u8>) -> Result {
+			ensure_signed(origin)?;
+			// TODO: https://github.com/paritytech/substrate/issues/1112
+			Ok(())
+		}
+
+		fn on_finalise(block_number: T::BlockNumber) {
+			if let Some(pending_change) = <PendingChange<T>>::get() {
+				if block_number == pending_change.scheduled_at {
+					Self::deposit_log(RawLog::AuthoritiesChangeSignal(
+						pending_change.delay,
+						pending_change.next_authorities.clone(),
+					));
+				}
+
+				if block_number == pending_change.scheduled_at + pending_change.delay {
+					Self::deposit_event(
+						RawEvent::NewAuthorities(pending_change.next_authorities.clone())
+					);
+					<AuthorityStorageVec<T::SessionKey>>::set_items(pending_change.next_authorities);
+					<PendingChange<T>>::kill();
+				}
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	/// Get the current set of authorities, along with their respective weights.
+	pub fn grandpa_authorities() -> Vec<(T::SessionKey, u64)> {
+		<AuthorityStorageVec<T::SessionKey>>::items()
+	}
+
+	/// Schedule a change in the authorities.
+	///
+	/// The change will be applied at the end of execution of the block
+	/// `in_blocks` after the current block. This value may be 0, in which
+	/// case the change is applied at the end of the current block.
+	///
+	/// No change should be signalled while any change is pending. Returns
+	/// an error if a change is already pending.
+	pub fn schedule_change(
+		next_authorities: Vec<(T::SessionKey, u64)>,
+		in_blocks: T::BlockNumber,
+	) -> Result {
+		if Self::pending_change().is_none() {
+			let scheduled_at = system::ChainContext::<T>::default().current_height();
+			<PendingChange<T>>::put(StoredPendingChange {
+				delay: in_blocks,
+				scheduled_at,
+				next_authorities,
+			});
+
+			Ok(())
+		} else {
+			Err("Attempt to signal GRANDPA change with one already pending.")
+		}
+	}
+
+	/// Deposit one of this module's logs.
+	fn deposit_log(log: Log<T>) {
+		<system::Module<T>>::deposit_log(<T as Trait>::Log::from(log).into());
+	}
+}
+
+impl<T: Trait> Module<T>
+where
+	AuthorityId: core::convert::From<<T as Trait>::SessionKey>
+{
+	/// See if the digest contains any scheduled change.
+	pub fn scrape_digest_change(log: &Log<T>)
+		-> Option<ScheduledChange<T::BlockNumber>>
+	{
+		<Log<T> as GrandpaChangeSignal<T::BlockNumber>>::as_signal(log)
+	}
+}
+
+/// Helper for authorities being synchronized with the general session authorities.
+///
+/// This is not the only way to manage an authority set for GRANDPA, but it is
+/// a convenient one. When this is used, no other mechanism for altering authority
+/// sets should be.
+pub struct SyncedAuthorities<T>(::rstd::marker::PhantomData<T>);
+
+impl<T> Default for SyncedAuthorities<T> {
```

Comment:
TODO: remove once the rust compiler guys fix derive: https://github.com/rust-lang/rust/issues/26925